File: ui/frontend/src/views/orderedit/order-details.js
import { PolymerElement } from '@polymer/polymer/polymer-element.js';
import '@polymer/iron-icon/iron-icon.js';
import '@vaadin/vaadin-icons/vaadin-icons.js';
import '@vaadin/vaadin-button/src/vaadin-button.js';
import '@vaadin/vaadin-form-layout/src/vaadin-form-item.js';
import '@vaadin/vaadin-form-layout/src/vaadin-form-layout.js';
import '@vaadin/vaadin-text-field/src/vaadin-text-field.js';
import '../../components/buttons-bar.js';
import '../../components/utils-mixin.js';
import '../storefront/order-status-badge.js';
import '../../../styles/shared-styles.js';
import { html } from '@polymer/polymer/lib/utils/html-tag.js';
class OrderDetails extends window.ScrollShadowMixin(PolymerElement) {
  static get template() {
    return html`
    <style include="shared-styles">
      :host {
        display: flex;
        flex-direction: column;
        box-sizing: border-box;
        flex: auto;
      }

      /*
        Workaround for non-working dom-repeat inside tables in IE11
        (https://github.com/Polymer/polymer/issues/1567):
        use divs with table-like display values instead of the actual
        <table>, <tr> and <td> elements.
      */
      .table {
        display: table;
      }

      .tr {
        display: table-row;
      }

      .td {
        display: table-cell;
      }

      .main-row {
        flex: 1;
      }

      h3 {
        margin: 0;
        word-break: break-all;
        /* Non standard for WebKit */
        word-break: break-word;
        white-space: normal;
      }

      .date,
      .time {
        white-space: nowrap;
      }

      .dim,
      .secondary {
        color: var(--lumo-secondary-text-color);
      }

      .secondary {
        font-size: var(--lumo-font-size-xs);
        line-height: var(--lumo-font-size-xl);
      }

      .meta-row {
        display: flex;
        justify-content: space-between;
        padding-bottom: var(--lumo-space-s);
      }

      .products {
        width: 100%;
      }

      .products .td {
        text-align: center;
        vertical-align: middle;
        padding: var(--lumo-space-xs);
        border: none;
        border-bottom: 1px solid var(--lumo-contrast-10pct);
      }

      .products .td.product-name {
        text-align: left;
        padding-left: 0;
        width: 100%;
      }

      .products .td.number {
        text-align: right;
      }

      .products .td.money {
        text-align: right;
        padding-right: 0;
      }

      .history-line {
        margin: var(--lumo-space-xs) 0;
      }

      .comment {
        font-size: var(--lumo-font-size-s);
      }

      order-status-badge[small] {
        margin-left: 0.5em;
      }

      #sendComment {
        color: var(--lumo-primary-color-50pct);
      }

      @media (min-width: 600px) {
        .main-row {
          padding: var(--lumo-space-l);
          flex-basis: auto;
        }
      }
    </style>

    <div class="scrollable main-row" id="main">
      <div class="meta-row">
        <order-status-badge status="[[item.state]]"></order-status-badge>
        <span class="dim">Order #[[item.id]]</span>
      </div>

      <vaadin-form-layout id="form1">
        <vaadin-form-item>
          <label slot="label">Due</label>
          <vaadin-form-layout id="form2">
            <div class="date">
              <h3>[[item.dueDate.day]]</h3>
              <span class="dim">[[item.dueDate.weekday]]</span>
            </div>
            <div class="time">
              <h3>[[item.dueTime]]</h3>
              <span class="dim">[[item.pickupLocation.name]]</span>
            </div>
          </vaadin-form-layout>
        </vaadin-form-item>

        <vaadin-form-item colspan="2">
          <label slot="label">Customer</label>
          <h3>[[item.customer.fullName]]</h3>
        </vaadin-form-item>

        <vaadin-form-item>
          <label slot="label">Phone number</label>
          <h3>[[item.customer.phoneNumber]]</h3>
        </vaadin-form-item>
      </vaadin-form-layout>

      <vaadin-form-layout id="form3">
        <div></div>

        <vaadin-form-layout id="form4" colspan="2">
          <template is="dom-if" if="[[item.customer.details]]">
            <vaadin-form-item label-position="top">
              <label slot="label">Additional details</label>
              <span>[[item.customer.details]]</span>
            </vaadin-form-item>
          </template>

          <vaadin-form-item>
            <label slot="label">Products</label>
            <div class="table products">
              <template is="dom-repeat" items="[[item.items]]" as="item">
                <dom-if if="[[item.product.name]]">
                  <template>
                    <div class="tr">
                      <div class="td product-name">
                        <div class="bold">[[item.product.name]]</div>
                        <div class="secondary">[[item.comment]]</div>
                      </div>
                      <div class="td number">
                        <span class="count">[[item.quantity]]</span>
                      </div>
                      <div class="td dim">×</div>
                      <div class="td money">
                        [[item.product.price]]
                      </div>
                    </div>
                  </template>
                </dom-if>
              </template>
            </div>
          </vaadin-form-item>

          <vaadin-form-item id="history" label-position="top" hidden="[[review]]">
            <label slot="label">History</label>
            <template is="dom-repeat" items="[[item.history]]" as="event">
              <div class="history-line">
                <span class="bold">[[event.createdBy.firstName]]</span>
                <span class="secondary">[[event.timestamp]]</span>
                <order-status-badge status="[[event.newState]]" small=""></order-status-badge>
              </div>
              <div class="comment">[[event.message]]</div>
            </template>
          </vaadin-form-item>

          <vaadin-form-item id="comment" hidden="[[review]]">
            <vaadin-text-field id="commentField" placeholder="Add comment" class="full-width" on-keydown="_onCommentKeydown" 
              maxlength="255">
              <div slot="suffix" class="comment-suffix">
                <vaadin-button id="sendComment" theme="tertiary">Send</vaadin-button>
              </div>
            </vaadin-text-field>
          </vaadin-form-item>

        </vaadin-form-layout>
      </vaadin-form-layout>
    </div>


    <buttons-bar id="footer" no-scroll\$="[[noScroll]]">
      <vaadin-button slot="left" id="back" hidden="[[!review]]">Back</vaadin-button>
      <vaadin-button slot="left" id="cancel" hidden="[[review]]">Cancel</vaadin-button>

      <div slot="info" class="total">Total [[item.totalPrice]]</div>

      <vaadin-button slot="right" id="save" theme="primary success" hidden="[[!review]]">
        <iron-icon icon="vaadin:check" slot="suffix"></iron-icon>
        Place order</vaadin-button>
      <vaadin-button slot="right" id="edit" theme="primary" hidden="[[review]]">
        Edit order
        <iron-icon icon="vaadin:edit" slot="suffix"></iron-icon>
      </vaadin-button>
    </buttons-bar>
`;
  }

  static get is() {
    return 'order-details';
  }

  static get properties() {
    return {
      item: {
        type: Object
      }
    };
  }

  ready() {
    super.ready();

    this.$.form1.responsiveSteps = this.$.form3.responsiveSteps = [
      {columns: 1, labelsPosition: 'top'},
      {minWidth: '600px', columns: 4, labelsPosition: 'top'}
    ];

    this.$.form2.responsiveSteps = [
      {columns: 1}, {minWidth: '180px', columns: 2}
    ];

    this.$.form4.responsiveSteps = [
      {columns: 1, labelsPosition: 'top'}
    ];
  }

  _onCommentKeydown(event) {
    if (event.key === 'Enter' || event.keyCode == 13) {
      // In IE11 on button click commentField blur doesn't happen, and the value-change event is not fired
      this.$.commentField.blur();
      this.$.sendComment.click();
    }
  }
}

window.customElements.define(OrderDetails.is, OrderDetails);

File: ui/src/main/java/com/marketcetera/colin/ui/views/orderedit/OrderDetails.java
/**
 *
 */
package com.marketcetera.colin.ui.views.orderedit;

import com.vaadin.flow.component.ClickEvent;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.polymertemplate.Id;
import com.vaadin.flow.component.polymertemplate.PolymerTemplate;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.dom.Element;
import com.vaadin.flow.shared.Registration;
import com.vaadin.flow.templatemodel.Encode;
import com.vaadin.flow.templatemodel.Include;
import com.vaadin.flow.templatemodel.TemplateModel;
import com.marketcetera.colin.backend.data.entity.Order;
import com.marketcetera.colin.ui.events.CancelEvent;
import com.marketcetera.colin.ui.events.SaveEvent;
import com.marketcetera.colin.ui.utils.converters.CurrencyFormatter;
import com.marketcetera.colin.ui.utils.converters.LocalDateTimeConverter;
import com.marketcetera.colin.ui.utils.converters.LocalTimeConverter;
import com.marketcetera.colin.ui.utils.converters.LongToStringConverter;
import com.marketcetera.colin.ui.utils.converters.OrderStateConverter;
import com.marketcetera.colin.ui.views.storefront.converters.StorefrontLocalDateConverter;
import com.marketcetera.colin.ui.views.storefront.events.CommentEvent;
import com.marketcetera.colin.ui.views.storefront.events.EditEvent;

/**
 * The component displaying a full (read-only) summary of an order, and a comment
 * field to add comments.
 */
@Tag("order-details")
@JsModule("./src/views/orderedit/order-details.js")
public class OrderDetails extends PolymerTemplate<OrderDetails.Model> {

	private Order order;

	@Id("back")
	private Button back;

	@Id("cancel")
	private Button cancel;

	@Id("save")
	private Button save;

	@Id("edit")
	private Button edit;

	@Id("history")
	private Element history;

	@Id("comment")
	private Element comment;

	@Id("sendComment")
	private Button sendComment;

	@Id("commentField")
	private TextField commentField;

	private boolean isDirty;

	public OrderDetails() {
		sendComment.addClickListener(e -> {
			String message = commentField.getValue();
			message = message == null ? "" : message.trim();
			if (!message.isEmpty()) {
				commentField.clear();
				fireEvent(new CommentEvent(this, order.getId(), message));
			}
		});
		save.addClickListener(e -> fireEvent(new SaveEvent(this, false)));
		cancel.addClickListener(e -> fireEvent(new CancelEvent(this, false)));
		edit.addClickListener(e -> fireEvent(new EditEvent(this)));
	}

	public void display(Order order, boolean review) {
		getModel().setReview(review);
		this.order = order;
		getModel().setItem(order);
		if (!review) {
			commentField.clear();
		}
		this.isDirty = review;
	}

	public boolean isDirty() {
		return isDirty;
	}

	public void setDirty(boolean isDirty) {
		this.isDirty = isDirty;
	}

	public interface Model extends TemplateModel {
		@Include({ "id", "dueDate.day", "dueDate.weekday", "dueDate.date", "dueTime", "state", "pickupLocation.name",
			"customer.fullName", "customer.phoneNumber", "customer.details", "items.product.name", "items.comment",
			"items.quantity", "items.product.price", "history.message", "history.createdBy.firstName",
			"history.timestamp", "history.newState", "totalPrice" })
		@Encode(value = LongToStringConverter.class, path = "id")
		@Encode(value = StorefrontLocalDateConverter.class, path = "dueDate")
		@Encode(value = LocalTimeConverter.class, path = "dueTime")
		@Encode(value = OrderStateConverter.class, path = "state")
		@Encode(value = CurrencyFormatter.class, path = "items.product.price")
		@Encode(value = LocalDateTimeConverter.class, path = "history.timestamp")
		@Encode(value = OrderStateConverter.class, path = "history.newState")
		@Encode(value = CurrencyFormatter.class, path = "totalPrice")
		void setItem(Order order);

		void setReview(boolean review);
	}

	public Registration addSaveListenter(ComponentEventListener<SaveEvent> listener) {
		return addListener(SaveEvent.class, listener);
	}

	public Registration addEditListener(ComponentEventListener<EditEvent> listener) {
		return addListener(EditEvent.class, listener);
	}

	public Registration addBackListener(ComponentEventListener<ClickEvent<Button>> listener) {
		return back.addClickListener(listener);
	}

	public Registration addCommentListener(ComponentEventListener<CommentEvent> listener) {
		return addListener(CommentEvent.class, listener);
	}

	public Registration addCancelListener(ComponentEventListener<CancelEvent> listener) {
		return addListener(CancelEvent.class, listener);
	}
}
